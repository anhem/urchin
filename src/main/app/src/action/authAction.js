import 'isomorphic-fetch';
import {get} from './restClient'
import {Actions} from "../constants";

const {Auth} = Actions;

export const login = (username, password) => (dispatch) => {
    dispatch({
        type: Auth.LOGIN
    });

    return fetch('/login', {
        method: 'post',
        headers: {
            'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8'
        },
        credentials: 'include',
        body: 'username=' + username + '&password=' + encodeURIComponent(password)
    }).then(response => {
        if (response.status >= 500) {
            dispatch({
                type: Auth.LOGIN_FAILED
            });
        } else if (response.status === 200) {
            dispatch({
                type: Auth.LOGIN_SUCCESS
            });
        } else {
            dispatch({
                type: Auth.LOGIN_INVALID
            });
        }
    }, error => {
        dispatch({
            type: Auth.LOGIN_FAILED,
            data: error
        });
    })
};

export const isAuthenticated = () => (dispatch) => {
    return get('/api/authenticate')
        .then(json => {
            dispatch({
                type: Auth.LOGIN_SUCCESS
            });
        }, error => {
            if (error.httpStatus === 428) {
                dispatch({
                    type: Auth.PRECONDITION_FAILED
                });
            } else {
                dispatch({
                    type: Auth.LOGOUT_SUCCESS
                });
            }
        })
};