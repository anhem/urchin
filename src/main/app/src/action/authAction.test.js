import thunk from 'redux-thunk'
import configureMockStore from 'redux-mock-store'
import fetchMock from 'fetch-mock'
import {isAuthenticated, login} from "./authAction";

describe('authAction', () => {

    const username = 'username';
    const password = 'password';

    let store;

    beforeEach(() => {
        store = configureMockStore([thunk])({});
    });

    afterEach(() => {
        fetchMock.reset();
        fetchMock.restore();
    });

    it('authenticate is successful', () => {
        fetchMock.getOnce('/api/authenticate', {
            message: 'ok'
        });

        return store.dispatch(isAuthenticated()).then(() => {
            expect(store.getActions()).toMatchSnapshot();
        })
    });

    it('authenticate failed', () => {
        return store.dispatch(isAuthenticated()).then(() => {
            expect(store.getActions()).toMatchSnapshot();
        })
    });

    it('authenticate precondition failed', () => {
        fetchMock.getOnce('/api/authenticate', {
            status: 428,
            body: {}
        });

        return store.dispatch(isAuthenticated()).then(() => {
            expect(store.getActions()).toMatchSnapshot();
        })
    });

    it('login is successful', () => {
        fetchMock.postOnce('/login', 200);

        return store.dispatch(login(username, password)).then(() => {
            expect(store.getActions()).toMatchSnapshot();
        })
    });

    it('login failed', () => {
        fetchMock.postOnce('/login', 500);

        return store.dispatch(login(username, password)).then(() => {
            expect(store.getActions()).toMatchSnapshot();
        })
    });

    it('login failed network error', () => {
        return store.dispatch(login(username, password)).then(() => {
            expect(store.getActions()).toMatchSnapshot();
        })
    });

    it('login invalid', () => {
        fetchMock.postOnce('/login', 400);

        return store.dispatch(login(username, password)).then(() => {
            expect(store.getActions()).toMatchSnapshot();
        })
    });

});