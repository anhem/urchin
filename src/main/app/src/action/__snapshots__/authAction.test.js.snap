// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authAction authenticate failed 1`] = `
Array [
  Object {
    "type": "LOGOUT_SUCCESS",
  },
]
`;

exports[`authAction authenticate is successful 1`] = `
Array [
  Object {
    "type": "LOGIN_SUCCESS",
  },
]
`;

exports[`authAction authenticate precondition failed 1`] = `
Array [
  Object {
    "type": "PRECONDITION_FAILED",
  },
]
`;

exports[`authAction login failed 1`] = `
Array [
  Object {
    "type": "LOGIN",
  },
  Object {
    "type": "LOGIN_FAILED",
  },
]
`;

exports[`authAction login failed network error 1`] = `
Array [
  Object {
    "type": "LOGIN",
  },
  Object {
    "data": [TypeError: Network request failed],
    "type": "LOGIN_FAILED",
  },
]
`;

exports[`authAction login invalid 1`] = `
Array [
  Object {
    "type": "LOGIN",
  },
  Object {
    "type": "LOGIN_INVALID",
  },
]
`;

exports[`authAction login is successful 1`] = `
Array [
  Object {
    "type": "LOGIN",
  },
  Object {
    "type": "LOGIN_SUCCESS",
  },
]
`;
